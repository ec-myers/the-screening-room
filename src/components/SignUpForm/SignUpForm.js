import React, { Component } from 'react';
import { Redirect } from 'react-router-dom';
import { createUser } from '../../apiCalls/apiCalls';
import { setUser, hasEmailError } from '../../actions/index';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import './SignUpForm.scss';

export class SignUpForm extends Component {
  constructor() {
    super();
    this.state = {
      name: '',
      email: '',
      password: '',
      isLoggedIn: false
    };
  }

  handleChange = event => {
    this.setState({
      [event.target.name]: event.target.value
    });
  };

  handleClick = async () => {
    const { setUser, hasEmailError } = this.props;
    try {
      let newUser = await createUser(this.state);
      setUser(newUser);
      this.setState({ isLoggedIn: true });
      hasEmailError('');
    } catch ({ message }) {
      hasEmailError(message);
    }
    this.clearInputs();
  };

  clearInputs = () => {
    const stateKeys = Object.keys(this.state);
    stateKeys.forEach(stateKey =>
      this.setState({
        [stateKey]: ''
      })
    );
  };

  handleSubmit = event => {
    event.preventDefault();
    this.handleClick();
  };

  render() {
    if (this.state.isLoggedIn) {
      return <Redirect to='/' />;
    }
    const { name, email, password } = this.state;
    const { emailErrMsg } = this.props;
    return (
      <form className='form-model' onSubmit={e => this.handleSubmit(e)}>
        {emailErrMsg ? (
          <p className='email-error signup-message'>{emailErrMsg}</p>
        ) : (
          <p className='signup-message'>
            Create an account with your email to keep track of your favorite
            movies.
          </p>
        )}
        <div>
          <label htmlFor='email' className='Form__label'>
            Your name
          </label>
          <input
            id='name'
            type='text'
            placeholder='Walter White'
            className='Form__input Form__input--name'
            name='name'
            value={name}
            onChange={e => this.handleChange(e)}
          ></input>
        </div>
        <div>
          <label htmlFor='email' className='Form__label'>
            Your email
          </label>
          {emailErrMsg ? (
            <input
              id='email'
              type='text'
              placeholder='your@email.com'
              className='Form__input--error Form__input'
              name='email'
              value={email}
              onChange={e => this.handleChange(e)}
            ></input>
          ) : (
            <input
              id='email'
              type='text'
              placeholder='your@email.com'
              className='Form__input Form__input--email'
              name='email'
              value={email}
              onChange={e => this.handleChange(e)}
            ></input>
          )}
        </div>
        <div>
          <label htmlFor='password' className='Form__label'>
            Your password
          </label>
          <input
            id='password'
            type='password'
            placeholder='Must have at least 8 characters'
            className='Form__input'
            name='password'
            value={password}
            onChange={e => this.handleChange(e)}
          ></input>
        </div>
        <button
          type='submit'
          className='Form__button signup-form__button'
        >
          Continue
        </button>
      </form>
    );
  }
}

export const mapStateToProps = ({ emailErrMsg, isLoading }) => ({
  emailErrMsg,
  isLoading
});

export const mapDispatchToProps = dispatch => {
  return bindActionCreators({ setUser, hasEmailError }, dispatch);
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(SignUpForm);
