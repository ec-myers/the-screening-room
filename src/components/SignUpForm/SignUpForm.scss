.signup-message {
  margin-bottom: 30px;
}