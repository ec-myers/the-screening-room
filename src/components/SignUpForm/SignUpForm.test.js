import React from 'react';
import { shallow } from 'enzyme';
import { SignUpForm, mapStateToProps, mapDispatchToProps } from './SignUpForm';

describe('SignUpForm', () => {
  let wrapper;
  const mockEvent = {
    target: {
      name: 'name',
      value: 'Vanessa Randall'
    }
  };
  const mockEvent2 = {
    target: {
      name: 'email',
      value: 'vanessa.randall@doane.edu'
    }
  };

  beforeEach(() => {
    wrapper = shallow(<SignUpForm emailErrMsg={''} isLoading={false} />);
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should update it's state property onChange of inputs", () => {
    wrapper.instance().handleChange(mockEvent);
    wrapper.instance().handleChange(mockEvent2);

    expect(wrapper.state()).toEqual({
      name: 'Vanessa Randall',
      email: 'vanessa.randall@doane.edu',
      password: '',
      isLoggedIn: false
    });
  });

  it('should call handleClick on submit', () => {
    wrapper.instance().handleSubmit = jest.fn();
    wrapper.instance().forceUpdate();
    wrapper.find('.form-model').simulate('submit');

    expect(wrapper.instance().handleSubmit).toHaveBeenCalled();
  });

  it('should invoke handleChange on change of email input', () => {
    let wrapper1 = shallow(
      <SignUpForm emailErrMsg={'Error'} isLoading={false} />
    );
    wrapper1.instance().forceUpdate();
    wrapper1.instance().handleChange = jest.fn();
    wrapper1.find('.Form__input--error').simulate('change', mockEvent);

    expect(wrapper1.instance().handleChange).toHaveBeenCalled();
  });

  it('should invoke handleChange on change of name input', () => {
    wrapper.instance().handleChange = jest.fn();
    wrapper.find('.Form__input--name').simulate('change', mockEvent);

    expect(wrapper.instance().handleChange).toHaveBeenCalled();
  });

  // it('should redirect the user to the homepage if the user is logged in', () => {
  //   wrapper.instance().forceUpdate();
  //   wrapper.instance().setState({
  //     isLoggedIn: true
  //   })

  // })

  it('should invoke handleChange on change of email input when emailErrMsg is falsey', () => {
    wrapper.instance().forceUpdate();
    const mockEvent = {
      target: {
        name: 'email',
        value: 'vanessa.randall@doane.edu'
      }
    };
    wrapper.instance().handleChange = jest.fn();
    wrapper.find('.Form__input--email').simulate('change', mockEvent);

    expect(wrapper.instance().handleChange).toHaveBeenCalled();
  });
});
