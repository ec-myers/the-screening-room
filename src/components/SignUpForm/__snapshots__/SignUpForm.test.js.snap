// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpForm should match the snapshot 1`] = `
<form
  className="form-model"
  onSubmit={[Function]}
>
  <p
    className="signup-message"
  >
    Create an account with your email to keep track of your favorite movies.
  </p>
  <div>
    <label
      className="Form__label"
      htmlFor="email"
    >
      Your name
    </label>
    <input
      className="Form__input Form__input--name"
      id="name"
      name="name"
      onChange={[Function]}
      placeholder="Walter White"
      type="text"
      value=""
    />
  </div>
  <div>
    <label
      className="Form__label"
      htmlFor="email"
    >
      Your email
    </label>
    <input
      className="Form__input Form__input--email"
      id="email"
      name="email"
      onChange={[Function]}
      placeholder="your@email.com"
      type="text"
      value=""
    />
  </div>
  <div>
    <label
      className="Form__label"
      htmlFor="password"
    >
      Your password
    </label>
    <input
      className="Form__input"
      id="password"
      name="password"
      onChange={[Function]}
      placeholder="Must have at least 8 characters"
      type="password"
      value=""
    />
  </div>
  <button
    className="Form__button signup-form__button"
    type="submit"
  >
    Continue
  </button>
</form>
`;
